Prefer content types from 'consumes' list in Open API spec

artem-smotrakov opened the pull request:
Open API specification defines  `consumes` parameter for a endpoint which allows to specify content types which the endpoint can consume. Many RESTful applications use standard content types such as `application/json`, but some of them define custom types like `application/vnd.awesome.application+json`. Such endpoints may respond with 415 error code (Unsupported Media Type) to requests which don't specify an expected custom content type in `Content-Type` header.

The Open API plugin is aware of `consumes` parameter but it usually prefers to use the standard `application/json` type for JSON payloads. As a result, it may be hard to test a endpoint which requires a custom content type. In fact, I couldn't configure w3af to use content types defined in `consumes`. As a workaround, we can try to set `Content-Type` header in `http-settings -> headers_file`. But if a RESTful application provides multiple endpoints which use different custom content types, then we can efficiently scan all of them in one scan.

In this patch, I'd like to propose updating w3af to prefer content types from `consumes` list.

Changes:

- Updated `RequestFactory` to prefer content types from `consumes` list
- Added `set_header()` method to `JSONContainer` to support setting headers (I noticed that headers from a data container may override the headers which were set to a fuzzable request before, see `FuzzableRequest.get_all_headers()` method)
- Updated `DataContainer` with default implementation of `set_header()` method
- Added several tests

artem-smotrakov committed:
Prefer content types from 'consumes' list in Open API spec

artem-smotrakov committed:
Added TestJSONContainer.test_headers_immutable() test

artem-smotrakov committed:
Added TestJSONContainer.test_pickle() test

andresriancho commented on w3af/core/data/parsers/doc/open_api/requests.py:
Note: If the API has `application/vnd.w3af` content-type (note that I removed the `+json` from your example content type) then [this line](https://github.com/andresriancho/w3af/blob/develop/w3af/core/data/parsers/doc/open_api/requests.py#L232) might set `dc` to None because that is [not one of the supported data container types](https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/core/data/dc/factory.py#L88-L108) , that happens because the content type has no `json` which doesn't match [here](https://github.com/andresriancho/w3af/blob/356b14b975039706f4fd7f4f5db5b114cd75f14e/w3af/core/data/dc/json_container.py#L70).

Not sure how common it is for APIs to define their custom content-type like this. Just mentioning it so you can take it into account.

artem-smotrakov replied:
Thanks for the comments. Makes sense to me, I'll try to update the code with the strategy you proposed. I think that w3af should throw an exception if it can't find a proper data container for an unknown content type like `application/vnd.w3af`.

andresriancho commented:
Thanks for another great contribution!

Added a few comments to the code, and here are extra questions:

 * When an API that uses `application/vnd.w3af+json` is parsed, are HTTP requests including that content-type? Did you check that with a proxy or logs? I saw multiple unit-tests, but none that verified that. I'm worried that at some point the header value could be overwritten.

artem-smotrakov commented:
> When an API that uses application/vnd.w3af+json is parsed, are HTTP requests including that content-type? Did you check that with a proxy or logs? I saw multiple unit-tests, but none that verified that. I'm worried that at some point the header value could be overwritten.

Yeah, I checked in the application logs that the custom header really reach the endpoints. But I usually use a config which instructs w3af to dump all HTTP requests to a file:

```
output config text_file
set output_file output-w3af.txt
set http_output_file output-http.txt
set verbose True
```

Hope it's a reliable way to check what is actually sent out :)

In fact, the first version of the patch didn't take into account that the headers may be overwritten in `FuzzableRequest.get_all_headers()` method. I noticed that by inspecting `output-http.txt`.

I tried to add a couple of test cases to `w3af/plugins/tests` but noticed that the test fail intermittently. I had a quick look at the failures, and looks like `httpretty` doesn't always work fine. Maybe `httpretty` can't work in a multi-threaded environment. Please let me know if you know a better way to test it.

andresriancho commented:
> Hope it's a reliable way to check what is actually sent out :)

Yes, in my experience that is reliable. If it was there, then we're fine on that point!

> Maybe httpretty can't work in a multi-threaded environment

https://github.com/gabrielfalcao/HTTPretty/pull/333 is something I sent to them a while ago, I believe it might be related.

At some point the [test requirements file](https://github.com/andresriancho/w3af/blob/develop/w3af/tests/requirements.txt#L34) was pointing to my httpretty fork. At this point I lost track of this issue and I'm not sure if they merged something to fix it upstream.

artem-smotrakov committed:
Look for the best content type for an API endpoint

artem-smotrakov commented:
@andresriancho I have updated the patch as you proposed. Please let me know if I missed something. A couple of notes:

- Added `XmlRpcContainer` to the list of data containers to be considered even if this data container is not currently used for scanning API endpoints. If this date container is updated in future, the plugin should be able to use it when it parses a `consumes` list.
- Even if `RequestFactory` throws an exception when no data container was found for content types from `consumes` list, the OpenAPI plugin will successfully return from `parse()` method because the method catches all exceptions [here](https://github.com/andresriancho/w3af/blob/develop/w3af/core/data/parsers/doc/open_api/main.py#L153). But the method won't return an API endpoint which it failed to process.
- I am not sure that `ValueError` is the best exception to throw in `get_consuming_content_type()` but not sure what would be a better option either.

artem-smotrakov committed:
Don't use XmlRpcContainer in get_consuming_content_type() method

artem-smotrakov commented on w3af/core/data/parsers/doc/open_api/requests.py:
```diff
@@ -228,6 +264,10 @@ def get_data_container(self, headers):
         # Create the data container
         dc = dc_from_content_type_and_raw_params(content_type, parameters)
+        if dc is None:
+            om.out.error("No data container for content type '%s'" % content_type)
```
Let me check.

artem-smotrakov committed:
Removed unused import in open_api/requests.py

artem-smotrakov commented:
Cool, I'll try to use the custom httpretty next time. Thanks for the review @andresriancho !
